R/pseudocolor_plot.R

```diff
@@ -0,0 +1,346 @@
+      marker2 %in% rownames(object),
+
+    "'grid_n' must be a numeric of length 1" =
+      is.numeric(grid_n),
+    "'scale_density' must be either TRUE or FALSE" =
+      is.logical(scale_density),
+    "'margin_density' must be either TRUE or FALSE" =
+      is.logical(margin_density),
+
+    "'pt.size' must be a numeric of length 1" =
+      is.numeric(pt.size),
+    "'alpha' must be a numeric of length 1" =
+      is.numeric(alpha),
+    "'layer' must be a character of length 1" =
+      is.character(layer),
+    "'coord.fixed' must be either TRUE or FALSE" =
+      is.logical(coord.fixed),
+
+    "'plot_gate' must be either NULL or a data.frame" =
+      is.null(plot_gate) | is.data.frame(plot_gate),
+    "'plot_gate' must have columns 'xmin', 'xmax', 'ymin', 'ymax'" =
+      is.null(plot_gate) | all(c("xmin", "xmax", "ymin", "ymax") %in% colnames(plot_gate)),
+
+    "'object' must be a Seurat object" =
+      inherits(object, "Seurat"),
+
+    "Marginal density is not supported when 'facet_vars' is not NULL" =
+      is.null(facet_vars) | !isTRUE(margin_density)
+  )
+
+  # Get data
+  plot_data <-
+    FetchData(object, vars = c(facet_vars, marker1, marker2),
+              layer = layer)
+
+  plot_data <-
+    plot_data %>%
+    rename(marker1 = all_of(marker1),
+           marker2 = all_of(marker2)) %>%
+    group_by_at(all_of(facet_vars)) %>%
+    mutate(dens = .get2Ddensity(marker1, marker2, n = grid_n, ...)) %>%
+    ungroup()
+
+  if(isTRUE(scale_density)) {
+    plot_data <-
+      plot_data %>%
+      group_by_at(all_of(facet_vars)) %>%
+      mutate(dens = dens / max(dens)) %>%
+      ungroup()
+  }
+
+  # Set plot theme
+  plot_theme <-
+    theme_bw()
+
+  if(is.null(facet_vars)) {
+
+    plot_theme <-
+      plot_theme +
+      theme(panel.grid = element_blank(),
+            panel.spacing = unit(0, "lines"),
+            plot.margin = unit(c(0, 0, 0, 0), "lines"))
+
+  } else {
+
+    plot_theme <-
+      plot_theme +
+      theme(panel.grid = element_blank())
+
+  }
+
+  plot_range <-
+    range(c(plot_data$marker1, plot_data$marker2))
+
+  if(!is.null(plot_gate)) {
+    plot_range <-
+      range(c(plot_data$marker1, plot_data$marker2,
+              plot_gate$xmin, plot_gate$xmax,
+              plot_gate$ymin, plot_gate$ymax))
+  }
+
+  # Make plot
+  plot <-
+    plot_data %>%
+    ggplot(aes(marker1, marker2, color = dens)) +
+    geom_hline(yintercept = 0, color = "gray") +
+    geom_vline(xintercept = 0, color = "gray") +
+    geom_point(size = pt.size,
+               alpha = alpha,
+               show.legend = F) +
+    scale_color_viridis_c(option = "H") +
+    scale_x_continuous(limits = plot_range) +
+    scale_y_continuous(limits = plot_range) +
+    plot_theme +
+    labs(x = marker1,
+         y = marker2)
+
+  if(isTRUE(coord.fixed) && isTRUE(margin_density)) {
+    warning("Fixed coordinates ('coord.fixed' = TRUE) is not supported when 'margin_density' is TRUE")
+  }
```

Comment:
This is just a warning, but should we put these kind of things earlier in the function when specifically are errors?